timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/LocalCorruptionDetector.java

```diff
@@ -31,46 +32,63 @@
 
     private final ScheduledExecutorService executor = PTExecutors.newSingleThreadScheduledExecutor(
             new NamedThreadFactory(CORRUPTION_DETECTOR_THREAD_PREFIX, true));
+
     private final LocalCorruptionHandler corruptionHandler;
+    private final PaxosLogHistoryProvider historyProvider;
 
     private volatile CorruptionStatus localCorruptionState = CorruptionStatus.HEALTHY;
+    private volatile CorruptionHealthReport localCorruptionReport = CorruptionHealthReport.defaultHealthyReport();
 
-    public static LocalCorruptionDetector create(List<TimeLockCorruptionNotifier> corruptionNotifiers) {
-        LocalCorruptionDetector localCorruptionDetector = new LocalCorruptionDetector(corruptionNotifiers);
+    public static LocalCorruptionDetector create(
+            PaxosLogHistoryProvider historyProvider, List<TimeLockCorruptionNotifier> corruptionNotifiers) {
+        LocalCorruptionDetector localCorruptionDetector =
+                new LocalCorruptionDetector(historyProvider, corruptionNotifiers);
 
         //        TODO(snanda) - uncomment when TL corruption detection goes live
         //        timeLockLocalCorruptionDetector.scheduleWithFixedDelay();
         return localCorruptionDetector;
     }
 
-    private LocalCorruptionDetector(List<TimeLockCorruptionNotifier> corruptionNotifiers) {
+    private LocalCorruptionDetector(
+            PaxosLogHistoryProvider historyProvider, List<TimeLockCorruptionNotifier> corruptionNotifiers) {
+        this.historyProvider = historyProvider;
         this.corruptionHandler = new LocalCorruptionHandler(corruptionNotifiers);
     }
 
     private void scheduleWithFixedDelay() {
         executor.scheduleWithFixedDelay(
                 () -> {
-                    if (detectedSignsOfCorruption()) {
-                        killTimeLock();
-                    }
+                    localCorruptionReport = analyzeHistoryAndBuildCorruptionHealthReport();
+                    processLocalHealthReport();
                 },
                 TIMELOCK_CORRUPTION_ANALYSIS_INTERVAL.getSeconds(),
                 TIMELOCK_CORRUPTION_ANALYSIS_INTERVAL.getSeconds(),
                 TimeUnit.SECONDS);
     }
 
-    private void killTimeLock() {
-        localCorruptionState = CorruptionStatus.DEFINITIVE_CORRUPTION_DETECTED_BY_LOCAL;
-        corruptionHandler.notifyRemoteServersOfCorruption();
+    private CorruptionHealthReport analyzeHistoryAndBuildCorruptionHealthReport() {
+        return HistoryAnalyzer.corruptionHealthReportForHistory(historyProvider.getHistory());
+    }
+
+    private void processLocalHealthReport() {
+        localCorruptionState = getLocalCorruptionState(localCorruptionReport);
+        if (localCorruptionState.shootTimeLock()) {
```

Comment:
as flagged elsewhere, this gets confusing, I think because it looks like a command, but it's not